src/sync.rs

```diff
@@ -0,0 +1,59 @@
+//! A module containing functions and utilities useful for synchronizing state.
+
+#![cfg_attr(not(all(target_vendor = "nintendo", target_env = "agb")), allow(unused_variables))]
+
+use crate::io::irq::{IrqEnableSetting, IME};
+
+mod locks;
+mod statics;
+
+pub use locks::*;
+pub use statics::*;
+
+/// Marks that a given pointer is read by volatile means without actually
+/// reading it.
+#[inline(always)]
+pub fn volatile_mark_ro<T>(val: *const T) {
```

Comment:
I'm kinda very suspicious that this even works myself.

The [documentation for `asm!`](<https://doc.rust-lang.org/beta/unstable-book/library-features/asm.html#options>) states that "by default, an inline assembly block is treated the same way as an external FFI function call with a custom calling convention: it may read/write memory, have observable side effects, etc." This should work well enough, since an external FFI call is about the right thing for representing e.g. what the DMA unit does. (i.e. reads from the input `*const` and writes out to the `*mut` pointer -- if in a concurrent way that's hard to represent in Rust's type system.)

The thing that makes me doubt myself a lot is that this code is basically the same as [`std::hint::black_box`](<https://doc.rust-lang.org/std/hint/fn.black_box.html>) pretty much, except it uses the `asm!` macro instead of `llvm_asm!`.

... which is concerning because the documentation for that function says "black_box is only (and can only be) provided on a 'best-effort' basis. The extent to which it can block optimisations may vary depending upon the platform and code-gen backend used. Programs cannot rely on black_box for correctness in any way.". I'm not sure if that's there for future-proofing or if there's a known way that LLVM can optimize across `black_box` in an unwanted way already.

We could always just use `compiler_fence` instead of having these functions, honestly, which will work for sure at the cost of emitting calls to the empty `__sync_synchronize` function until that bug is fixed.

Reply:
I know this fixes the bug in the `hello_world` example in practice, but I don't know if it's guaranteed to always do so in the future.